freud/box.pyx

```diff
@@ -422,6 +422,104 @@ cdef class Box:
+            points_indices (:math:`\left(N, \right)` :class:`numpy.ndarray`):
+                Array of indices that correspond to a set of points, of same length and query_points_indices.
+
+        Returns:
+            :math:`\left(N, \right)` :class:`numpy.ndarray`:
+                Array of distances between query points and points.
+        """   # noqa: E501
+        query_points = np.asarray(query_points)
+        points = np.asarray(points)
+        query_point_indices = np.atleast_1d(query_point_indices)
+        point_indices = np.atleast_1d(point_indices)
+
+        flatten = query_point_indices.ndim == 1
+        query_points = np.atleast_2d(query_points)
+        points = np.atleast_2d(points)
+
+        query_points = freud.util._convert_array(query_points, shape=(None, 3))
+        points = freud.util._convert_array(points, shape=(None, 3))
+
+        query_point_indices = freud.util._convert_array(
+            query_point_indices, shape=(None,), dtype=np.uint32)
+        point_indices = freud.util._convert_array(
+            point_indices, shape=(None,), dtype=np.uint32)
+
+        cdef:
+            const float[:, ::1] l_query_points = query_points
+            const float[:, ::1] l_points = points
+            const unsigned int[::1] l_query_point_indices = \
+                query_point_indices
+            const unsigned int[::1] l_point_indices = point_indices
+            size_t n_query_points = query_points.shape[0]
+            size_t n_points = points.shape[0]
+            size_t n_query_indices = query_point_indices.shape[0]
+            size_t n_point_indices = point_indices.shape[0]
+            float[::1] dist = np.empty(
+                query_point_indices.shape[0], dtype=np.float32)
+
+        self.thisptr.computeDistances(
+            <vec3[float]*> &l_query_points[0, 0],
+            <vec3[float]*> &l_points[0, 0],
+            &l_query_point_indices[0], &l_point_indices[0],
+            <float *> &dist[0], n_query_points, n_points,
+            n_query_indices, n_point_indices
+        )
+
+        return np.squeeze(dist) if flatten else np.asarray(dist)
+
+    def compute_all_distances(self, query_points, points):
+        R"""Calculate distances between a set of points and all query points,
+        with periodic boundary conditions.
+
+        Args:
+            query_points (:math:`\left(N_{query_points}, 3 \right)` :class:`numpy.ndarray`):
+                Array of query points.
+            points (:math:`\left(N_{points}, 3 \right)` :class:`numpy.ndarray`):
+                Array of points with same length as query_points.
+
+        Returns:
+            :math:`\left(N_{points}, N_{query_points}, \right)` :class:`numpy.ndarray`:
+                Array of distances between corresponding query points and points.
+        """  # noqa: E501
+        query_points = np.asarray(query_points)
+        points = np.asarray(points)
+
+        flatten = query_points.ndim == 1
```

Comment:
This case is a bit different from `compute_distances` (where I agree `flatten` was unnecessary). If I provide a single query point (i.e. `[0, 0, 0]`) and a set of points (ie `[[0, 1, 0], [2, 0, 0]]`), I would expect something like `distances=[1, 2]`, rather than `distances=[[1, 2]]`. If it's best to maintain the 2D arrays in this case as well, I can remove `flatten` for `compute_all_distances` too.

Reply:
@klywang A couple considerations:
- The behavior should be symmetric (same with one query point and multiple points as with one point and multiple query points).
- What should happen with one point and one query point? A scalar? A 1D array with one element?

Reply:
I'd be OK with always returning a matrix even for (1xn), (mx1), or (1x1) outputs. However, I agree that if we force 2D now we can generalize later without breaking APIs, so if we can't agree now then it's better to require the stricter case.